py/ringbuf.c

```diff
@@ -27,89 +27,91 @@
 
 #include "ringbuf.h"
 
-bool ringbuf_init(ringbuf_t *r, uint8_t *buf, size_t capacity) {
+bool ringbuf_init(ringbuf_t *r, uint8_t *buf, size_t size) {
     r->heap = false;
     r->buf = buf;
-    r->size = capacity;
-    r->iget = r->iput = 0;
+    r->size = size;
+    r->used = 0;
+    r->next_read = 0;
+    r->next_write = 0;
     return r->buf != NULL;
 }
 
 // Dynamic initialization. This should be accessible from a root pointer.
-// capacity is the number of bytes the ring buffer can hold. The actual
-// size of the buffer is one greater than that, due to how the buffer
-// handles empty and full statuses.
-bool ringbuf_alloc(ringbuf_t *r, size_t capacity, bool long_lived) {
+bool ringbuf_alloc(ringbuf_t *r, size_t size, bool long_lived) {
+    bool result = ringbuf_init(r, gc_alloc(size, false, long_lived), size);
     r->heap = true;
-    r->buf = gc_alloc(capacity + 1, false, long_lived);
-    r->size = capacity + 1;
-    r->iget = r->iput = 0;
-    return r->buf != NULL;
+    return result;
 }
 
 void ringbuf_free(ringbuf_t *r) {
-    if (r->heap) {
+    if (r->heap && gc_alloc_possible()) {
+        // Only free if on the heap, and there is really a heap right now.
         gc_free(r->buf);
     }
     r->buf = (uint8_t *)NULL;
+    r->heap = false;
     r->size = 0;
+
     ringbuf_clear(r);
 }
 
 size_t ringbuf_capacity(ringbuf_t *r) {
-    return r->size - 1;
+    return r->size;
 }
 
-// Returns -1 if buffer is empty, else returns byte fetched.
+// Return -1 if buffer is empty, else return byte fetched.
 int ringbuf_get(ringbuf_t *r) {
-    if (r->iget == r->iput) {
+    if (r->used < 1) {
         return -1;
     }
-    uint8_t v = r->buf[r->iget++];
-    if (r->iget >= r->size) {
-        r->iget = 0;
+    uint8_t v = r->buf[r->next_read];
+    r->next_read++;
+    if (r->next_read >= r->size) {
+        r->next_read = 0;
     }
+    r->used--;
```

Comment:
Now, r->used is modified from both `get` and `put`. If either one potentially runs from an interrupt context, the modifications need to be atomic or done with interrupts disabled.

Reply:
In Discord, we discussed that ringbuf is NOT interrupt safe and removals are guarded by an interrupt disable where necessary. it would be good if this was commented, but not critical.